packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -959,7 +934,124 @@ where
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id
+                ),
+                inserted_in_historic AS (
+                    -- We immediately put this deleted entity into the historic table.
+                    -- As this should be done in a transaction, we should be safe that this move
+                    -- doesn't produce invalid state.
+                    INSERT INTO entity_histories(
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id, 
+                        archived
+                    )
+                    SELECT
+                        entity_id, version,
+                        entity_type_version_id,
+                        properties,
+                        left_order, right_order,
+                        left_entity_id, right_entity_id,
+                        owned_by_id, created_by_id, updated_by_id,
+                        $2::boolean
+                    FROM to_move_to_historic
+                    -- For metadata purposes, we return the entire entity
+                    RETURNING *
+                )
+                SELECT
+                    entity_id, inserted_in_historic.version,
+                    base_uri, type_ids.version,
+                    properties, -- Unused
+                    left_order, right_order,
+                    left_entity_id, right_entity_id,
+                    owned_by_id, created_by_id, updated_by_id
+                FROM inserted_in_historic
+                INNER JOIN type_ids ON inserted_in_historic.entity_type_version_id = type_ids.version_id;
+                "#,
+                &[
+                    &entity_id,
+                    &archived,
+                ],
+            )
+            .await
+            .into_report()
+            .change_context(InsertionError)?;
+
+        let link_metadata = match (
+            historic_entity.get(5),
+            historic_entity.get(6),
+            historic_entity.get(7),
+            historic_entity.get(8),
+        ) {
+            (left_order, right_order, Some(left_entity_id), Some(right_entity_id)) => Some(
+                LinkEntityMetadata::new(left_entity_id, right_entity_id, left_order, right_order),
+            ),
+            _ => None,
+        };
+
+        let base_uri = BaseUri::new(historic_entity.get(2))
+            .into_report()
+            .change_context(InsertionError)?;
+        let entity_type_id = VersionedUri::new(base_uri, historic_entity.get::<_, i64>(3) as u32);
```

Comment:
Not that I know of. The other place we do this cast we let the type be inferred and then cast later https://github.com/hashintel/hash/blob/328e2c917b8c6ddf23fb6e3957c77842239434c8/packages/graph/hash_graph/lib/graph/src/store/postgres/context/ontology.rs#L106

Reply:
Does it make sense to make one? Would it be helpful for query compilation @TimDiekmann?

Reply:
We would require two type aliases. One for the Postgres representation (`i64`) and one for our model (`u32`) which would be even more confusing. For query compilation, I use `UnsignedInteger` for filters and `SignedInteger` for parameters.

Reply:
I'm just always a little concerned about the generics and casting when we're querying from SQL. It's so dependent on ordering and it's also very easy to mess up.

Would it be bad to have a `type OntologyVersion = u32` and `type OntologyVersionPostgres = i64` and similar for other things including date-timestamps, etc.?

Perhaps we could do it as part of https://app.asana.com/0/1201095311341924/1203214689883091/f

Reply:
I think it would be good to have the `OntologyVersion` type be explicit, but not sure about the postgres one as `i64` would be the right type to use there. Maybe it would be ideal if we could implement `FromSql` on a custom type to not have to do the casting at all? `OntologyVersion` could be a newtype then, and we would just implement ToSql somewhere in the PG mod

Reply:
If implementing the trait would work and still give us the casting check and such in debug builds then I think it sounds sensible :)

Will link this convo in the asana task